File: backend/user_service/src/utils/circuit_breaker.go
package utils

import (
	"errors"
	"time"
)

type ReachedMaxOpenError struct {
	msg string
}

func (e *ReachedMaxOpenError) Error() string { return e.msg }

type CircuitBreakerState int

const (
	closed CircuitBreakerState = iota
	open
	half_open
)

type CircuitBreaker struct {
	failures      int
	successes     int
	openEnters    int
	TimeInterval  time.Duration
	lastOpenStart time.Time
	MaxFailures   int
	MinSuccesses  int
	MaxOpenEnters int
	state         CircuitBreakerState
}

func (cb *CircuitBreaker) BeforeRequest() bool {
	if cb.state == open {
		if time.Since(cb.lastOpenStart) > cb.TimeInterval {
			cb.state = half_open
			return true
		}
		return false
	}
	return true
}

func (cb *CircuitBreaker) Execute(f func() (any, error)) (any, error) {
	if cb.BeforeRequest() {
		res, err := f()
		err = cb.afterRequest(err)
		return res, err
	}
	return nil, errors.New("cannot connect")
}

func (cb *CircuitBreaker) ConnectRetry(f func() (any, error), retries int, timeout time.Duration) (any, error) {
	for i := 0; i <= retries; i++ {
		res, err := cb.Execute(f)
		if errors.Is(err, &ReachedMaxOpenError{}) {
			return nil, err
		} else if err != nil {
			time.Sleep(timeout)
		} else {
			return res, err
		}
	}
	return nil, errors.New("cannot connect")
}

func (cb *CircuitBreaker) swithToOpen() error {
	cb.state = open
	cb.lastOpenStart = time.Now()
	cb.openEnters += 1
	if cb.openEnters > cb.MaxOpenEnters {
		return &ReachedMaxOpenError{"max open reached"}
	}
	return errors.New("cannot connect")
}

func (cb *CircuitBreaker) switchToClosed() {
	cb.state = closed
	cb.failures = 0
	cb.openEnters = 0
}

func (cb *CircuitBreaker) afterRequest(err error) error {
	if err != nil {
		switch cb.state {
		case closed:
			cb.failures += 1
			if cb.failures >= cb.MaxFailures {
				return cb.swithToOpen()
			}
			return errors.New("cannot connect")
		case half_open:
			return cb.swithToOpen()
		}
	} else {
		switch cb.state {
		case closed:
			cb.failures = 0
		case half_open:
			cb.successes += 1
			if cb.successes >= cb.MinSuccesses {
				cb.switchToClosed()
			}
		}
		return nil
	}
	return errors.New("invalid state")
}

File: backend/scraper_service/utils/circuit_breaker.py
import time
from enum import Enum
import asyncio
import logging

class ReachedMaxOpenException(Exception):
    msg = 'max open reached'

class CircuitBreakerState(Enum):
    CLOSED = 0
    OPEN = 1
    HALF_OPEN = 2


class CircuitBreaker:
    def __init__(self, timeInterval, maxFailures, minSuccesses, maxOpenEnters) -> None:
        self._failures = 0
        self._successes = 0
        self._openEnters = 0
        self.TimeInterval = timeInterval
        self._lastOpenStart = 0
        self.maxFailures = maxFailures
        self.minSuccesses = minSuccesses
        self.maxOpenEnters = maxOpenEnters
        self._state = CircuitBreakerState.CLOSED

    def _before_request(self) -> bool:
        if self._state == CircuitBreakerState.OPEN:
            if time.time() - self._lastOpenStart > self.TimeInterval:
                self._state = CircuitBreakerState.HALF_OPEN
                return True
            return False
        return True
    
    async def execute(self, f: callable) -> any:
        if self._before_request():
            try:
                res = await f()
                self._after_request(True)
                return res
            except Exception as e:
                self._after_request(False)
        raise ConnectionError('cannot connect')
    
    async def connect_retry(self, f: callable, retries: int, timeout: int) -> any:
        for _ in range(retries):
            try:
                return await self.execute(f)
            except ReachedMaxOpenException as err:
                raise ReachedMaxOpenException
            except Exception as e:
                await asyncio.sleep(timeout)
        raise ConnectionError('cannot connect')
    
    def _switch_to_open(self):
        self._state = CircuitBreakerState.OPEN
        self._lastOpenStart = time.time()
        self._openEnters += 1
        if self._openEnters > self.maxOpenEnters:
            raise ReachedMaxOpenException
        raise ConnectionError('cannot connect')
    
    def _switch_to_closed(self):
        self._state = CircuitBreakerState.CLOSED
        self._failures, self._openEnters = 0, 0

    def _after_request(self, is_error: bool):
        if not is_error:
            if self._state == CircuitBreakerState.CLOSED:
                self._failures += 1
                if self._failures >= self.maxFailures:
                    return self._switch_to_open()
                raise ConnectionError('cannot connect')
            else:
                return self._switch_to_open()
        else:
            if self._state == CircuitBreakerState.CLOSED:
                self._failures = 0
            else:
                self._successes += 1
                if self._successes >= self.minSuccesses:
                    self._switch_to_closed()

